src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -21,11 +22,413 @@
 #include "distributed/foreign_constraint.h"
 #include "distributed/master_protocol.h"
 #include "distributed/multi_join_order.h"
+#include "nodes/pg_list.h"
+#include "utils/catcache.h"
 #include "utils/fmgroids.h"
 #include "utils/rel.h"
 #include "utils/relcache.h"
 #include "utils/ruleutils.h"
 #include "utils/syscache.h"
+#include "utils/lsyscache.h"
+
+typedef struct FRelGraph
+{
+	HTAB *nodeMap;
+	int nodeCount;
+	Oid *indexToOidArray;
+	bool **transitivityMatrix;
+}FRelGraph;
+
+typedef struct FRelNode
+{
+	Oid relationId;
+	uint32 index;
+	List *adjacencyList;
+}FRelNode;
+
+static FRelGraph *frelGraph = NULL;
+
+static void CreateTransitivityMatrix(void);
+static void InitializeIndexToOidMapping(void);
+
+/* this function is only exported in the regression tests */
+PG_FUNCTION_INFO_V1(get_foreign_key_relation);
+
+/*
+ * get_foreign_key_relation returns the list of table oids that is referenced
+ * by given oid recursively. It uses the foreign key relation graph if it exists
+ * in the cache, otherwise forms it up once. Second argument, which is a boolean,
+ * set the direction of reference. True means that the function will return the
+ * Oids of relations referenced by given Oid. False means the opposite direction.
+ */
+Datum
+get_foreign_key_relation(PG_FUNCTION_ARGS)
+{
+	FuncCallContext *functionContext = NULL;
+	ListCell *foreignRelationCell = NULL;
+
+	CheckCitusVersion(ERROR);
+
+	/* for the first we call this UDF, we need to populate the result to return set */
+	if (SRF_IS_FIRSTCALL())
+	{
+		Oid relationId = PG_GETARG_OID(0);
+		bool isReferencing = PG_GETARG_BOOL(1);
+		List *refList = GetForeignKeyRelation(relationId, isReferencing);
+
+		/* create a function context for cross-call persistence */
+		functionContext = SRF_FIRSTCALL_INIT();
+
+		foreignRelationCell = list_head(refList);
+		functionContext->user_fctx = foreignRelationCell;
+	}
+
+	/*
+	 * On every call to this function, we get the current position in the
+	 * statement list. We then iterate to the next position in the list and
+	 * return the current statement, if we have not yet reached the end of
+	 * list.
+	 */
+	functionContext = SRF_PERCALL_SETUP();
+
+	foreignRelationCell = (ListCell *) functionContext->user_fctx;
+	if (foreignRelationCell != NULL)
+	{
+		Oid refId = lfirst_oid(foreignRelationCell);
+
+		functionContext->user_fctx = lnext(foreignRelationCell);
+
+		SRF_RETURN_NEXT(functionContext, PointerGetDatum(refId));
+	}
+	else
+	{
+		SRF_RETURN_DONE(functionContext);
+	}
+}
+
+
+/*
+ * CreateForeignKeyRelationGraph creates the foreign key relation graph using
+ * foreign constraint provided by pg_constraint metadata table.
+ */
+void
+CreateForeignKeyRelationGraph()
```

Comment:
We also need an API to destroy the graph and release all the allocated memory. Could you please add that? Thanks!